File: src/core/Background.lua
--- src/visuals/Background.lua
-- @class Background
-- Background system with parallax effect
-- Implements a 4-layer parallax background where layer 1 is closest and layer 4 is farthest
local Background = {}

-- Table to store all background layers
Background.layers = {}

--- Load background layers
-- @param layerPaths Table of image paths for each layer (from closest to farthest)
-- @param parallaxFactors Table of parallax factors for each layer (smaller values = slower movement)
function Background:load(layerPaths, parallaxFactors)
    self.layers = {}

    -- Load each background layer
    for i, path in ipairs(layerPaths) do
        table.insert(self.layers, {
            image = love.graphics.newImage(path),
            parallaxFactor = parallaxFactors[i] or (1 - (i - 1) * 0.2), -- Default decreasing factors if not provided
            x = 0,
            y = 0, -- Allow for vertical positioning
        })
    end

    -- Get screen dimensions
    self.screenWidth = love.graphics.getWidth()
    self.screenHeight = love.graphics.getHeight()

    -- Position layers vertically to align with the bottom of the screen
    -- This ensures backgrounds with different heights align properly
    for _, layer in ipairs(self.layers) do
        local imgHeight = layer.image:getHeight()
        -- Position the image so it aligns with the bottom of the screen
        layer.y = self.screenHeight - imgHeight
    end
end

--- Update background positions based on camera movement
-- @param dt Delta time
-- @param cameraX Camera X position
function Background:update(dt, cameraX)
    for _, layer in ipairs(self.layers) do
        -- Update layer position based on camera and parallax factor
        -- The lower the parallax factor, the slower the layer moves
        layer.x = -cameraX * layer.parallaxFactor
    end
end

--- Draw all background layers
function Background:draw()
    -- Draw from farthest to closest (reverse order from our layers table)
    for i = #self.layers, 1, -1 do
        local layer = self.layers[i]
        local img = layer.image
        local imgWidth = img:getWidth()

        -- Calculate how many copies we need to cover the screen
        local copies = math.ceil(self.screenWidth / imgWidth) + 1

        -- Calculate the starting position for tiling
        local startX = math.floor(layer.x % imgWidth) - imgWidth

        -- Draw enough copies to cover the screen
        for j = 0, copies do
            love.graphics.draw(img, startX + (j * imgWidth), layer.y)
        end
    end
end

return Background

File: internal/core/background.go
package core

import (
	"image"
	_ "image/png" // register PNG decoder
	"math"
	"os"

	"github.com/hajimehoshi/ebiten/v2"
)

// layer is a single parallax background layer.
type layer struct {
	image          *ebiten.Image
	parallaxFactor float64
	x              float64
	y              float64
}

// Background manages multi-layer parallax scrolling.
type Background struct {
	layers       []layer
	screenWidth  int
	screenHeight int
}

// NewBackground loads background layers from the given image paths with corresponding
// parallax factors. Layers are ordered from closest (index 0) to farthest.
func NewBackground(layerPaths []string, parallaxFactors []float64, screenWidth, screenHeight int) (*Background, error) {
	bg := &Background{
		screenWidth:  screenWidth,
		screenHeight: screenHeight,
	}

	for i, path := range layerPaths {
		img, err := loadImage(path)
		if err != nil {
			return nil, err
		}

		factor := 0.5
		if i < len(parallaxFactors) {
			factor = parallaxFactors[i]
		}

		// Align layer to the bottom of the screen
		imgHeight := img.Bounds().Dy()
		yOffset := float64(screenHeight) - float64(imgHeight)

		bg.layers = append(bg.layers, layer{
			image:          img,
			parallaxFactor: factor,
			x:              0,
			y:              yOffset,
		})
	}

	return bg, nil
}

// Update recalculates layer positions based on camera X position.
func (bg *Background) Update(cameraX float64) {
	for i := range bg.layers {
		bg.layers[i].x = -cameraX * bg.layers[i].parallaxFactor
	}
}

// Draw renders all background layers from farthest to closest with seamless tiling.
func (bg *Background) Draw(screen *ebiten.Image) {
	// Draw from farthest (last) to closest (first)
	for i := len(bg.layers) - 1; i >= 0; i-- {
		l := &bg.layers[i]
		img := l.image
		imgWidth := img.Bounds().Dx()

		// Calculate how many copies needed to cover the screen
		copies := bg.screenWidth/imgWidth + 2

		// Calculate starting X position for seamless tiling
		startX := math.Floor(math.Mod(l.x, float64(imgWidth))) - float64(imgWidth)

		for j := 0; j <= copies; j++ {
			opts := &ebiten.DrawImageOptions{}
			opts.GeoM.Translate(startX+float64(j*imgWidth), l.y)
			screen.DrawImage(img, opts)
		}
	}
}

// loadImage loads a PNG file from disk as an *ebiten.Image.
func loadImage(path string) (*ebiten.Image, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	img, _, err := image.Decode(f)
	if err != nil {
		return nil, err
	}

	return ebiten.NewImageFromImage(img), nil
}
